File: src/main/scala/com/ajjpj/amapper/javabean/builder/JavaBeanMapperBuilder.scala
package com.ajjpj.amapper.javabean.builder

import com.ajjpj.amapper.AMapper
import com.ajjpj.amapper.core.impl.AMapperImpl
import com.ajjpj.amapper.core._
import com.ajjpj.amapper.javabean.{AnnotationBasedContextExtractor, BuiltinValueMappingDefs, SimpleBeanMappingHelper, JavaBeanMappingHelper}

/**
 * @author arno
 */
class JavaBeanMapperBuilder[H <: JavaBeanMappingHelper](val helperFactory: () => H) {
  import BuiltinValueMappingDefs._

  var valueMappings: List[AValueMappingDef[_,_,_>:H]] = List(
    StringMappingDef, BooleanMappingDef, DateMappingDef, ClassMappingDef, CharMappingDef, BigDecimalMappingDef, EnumMappingDef,
    ByteMappingDef, ShortMappingDef, IntMappingDef, LongMappingDef, FloatMappingDef, DoubleMappingDef
  )

  var objectMappings: List[AObjectMappingDef[_,_,_>:H]] = List()
  var log: AMapperLogger = AMapperLogger.defaultLogger

  var preProcessor = List[APreProcessor]()
  var postProcessor = List[APostProcessor]()

  var identifierExtractor: IdentifierExtractor = new IdentifierExtractor {
    def uniqueIdentifier(o: AnyRef, tpe: AType) = o.toString
  }
  var contextExtractor: AContextExtractor = AnnotationBasedContextExtractor

  /**
   * later additions take precedence over earlier additions
   */
  def addValueMapping(v: AValueMappingDef[_,_,_>:H]) = {valueMappings = v :: valueMappings; this}

  /**
   * later additions take precedence over earlier additions
   */
  def addObjectMapping(m: AObjectMappingDef[_,_,_>:H]): JavaBeanMapperBuilder[H] = {objectMappings = m :: objectMappings; this}

  def addBeanMapping(m: JavaBeanMapping[_,_]): JavaBeanMapperBuilder[H] = {
    addObjectMapping(m.build)
    if(m.sourceCls != m.targetCls)
      addObjectMapping(m.buildBackward)
    this
  }

  def withLogger(log: AMapperLogger): JavaBeanMapperBuilder[H] = {this.log = log; this}
  def withIdentifierExtractor(identifierExtractor: IdentifierExtractor) = {this.identifierExtractor = identifierExtractor; this}
  def withContextExtractor(contextExtractor: AContextExtractor) = {this.contextExtractor = contextExtractor; this}

  def build: AMapper = new AMapperImpl[H] (new CanHandleSourceAndTargetCache(valueMappings), new CanHandleSourceAndTargetCache(objectMappings), log, helperFactory,
    identifierExtractor, contextExtractor,
    new CanHandleSourceAndTargetCache (preProcessor), new CanHandleSourceAndTargetCache (postProcessor))
}

object JavaBeanMapperBuilder { //TODO remove this
  /**
   * Factory method that allows type inference from Java code
   */
  def create() = new JavaBeanMapperBuilder (() => SimpleBeanMappingHelper)

  /**
   * Factory method that allows type inference from Java code
   */
  def create[H <: JavaBeanMappingHelper] (helperFactory: () => H) = new JavaBeanMapperBuilder (helperFactory)
}

File: src/main/java/com/ajjpj/amapper/javabean/builder/JavaBeanMapperBuilder.java
package com.ajjpj.amapper.javabean.builder;

import com.ajjpj.abase.function.AFunction0;
import com.ajjpj.amapper.core.*;
import com.ajjpj.amapper.core.exclog.AMapperExceptionHandler;
import com.ajjpj.amapper.core.exclog.AMapperLogger;
import com.ajjpj.amapper.core.path.APath;
import com.ajjpj.amapper.core.tpe.AQualifiedType;
import com.ajjpj.amapper.javabean.AnnotationBasedContextExtractor;
import com.ajjpj.amapper.javabean.JavaBeanMapper;
import com.ajjpj.amapper.javabean.JavaBeanMappingHelper;
import com.ajjpj.amapper.javabean.SimpleJavaBeanMappingHelper;
import com.ajjpj.amapper.javabean.impl.JavaBeanMapperImpl;
import com.ajjpj.amapper.javabean.mappingdef.BuiltinValueMappingDefs;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * This builder is the typical starting point for creating an AMapper instance for Java Beans. It builds an instance
 *  of JavaBeanMapper, which is in turn a Java Bean friendly API wrapper around an AMapper instance.
 *
 * @author arno
 */
public class JavaBeanMapperBuilder <H extends JavaBeanMappingHelper> {
    private final List<AObjectMappingDef<?,?,? super H>> objectMappings = new ArrayList<AObjectMappingDef<?, ?, ? super H>>();
    private final List<AValueMappingDef <?,?,? super H>> valueMappings  = new ArrayList<AValueMappingDef <?, ?, ? super H>>(Arrays.asList(
            BuiltinValueMappingDefs.StringMappingDef, BuiltinValueMappingDefs.BooleanMappingDef, BuiltinValueMappingDefs.CharacterMappingDef,
            BuiltinValueMappingDefs.BigDecimalMappingDef, BuiltinValueMappingDefs.BigIntegerMappingDef,
            BuiltinValueMappingDefs.DateMappingDef, BuiltinValueMappingDefs.LocaleMappingDef, BuiltinValueMappingDefs.TimeZoneMappingDef, BuiltinValueMappingDefs.CurrencyMappingDef,
            BuiltinValueMappingDefs.UuidMappingDef,
            BuiltinValueMappingDefs.ClassMappingDef, BuiltinValueMappingDefs.EnumMappingDef,
            BuiltinValueMappingDefs.ByteMappingDef, BuiltinValueMappingDefs.ShortMappingDef, BuiltinValueMappingDefs.IntegerMappingDef, BuiltinValueMappingDefs.LongMappingDef,
            BuiltinValueMappingDefs.FloatMappingDef, BuiltinValueMappingDefs.DoubleMappingDef
    ));

    private AMapperLogger logger = AMapperLogger.defaultLogger();
    private AFunction0<H, RuntimeException> helperFactory = new AFunction0<H, RuntimeException>() {
        @SuppressWarnings("unchecked")
        @Override public H apply() {
            return (H) SimpleJavaBeanMappingHelper.INSTANCE;
        }
    };
    private AIdentifierExtractor identifierExtractor = new AIdentifierExtractor() {
        @Override
        public Object uniqueIdentifier(Object o, AQualifiedType type, AQualifiedType targetType) {
            return o == null ? null : o.toString();
        }
    };
    private AContextExtractor contextExtractor = new AnnotationBasedContextExtractor();

    private final List<APreProcessor> preProcessors = new ArrayList<APreProcessor>();
    private final List<APostProcessor> postProcessors = new ArrayList<APostProcessor>();

    //---------------------------------

    public static JavaBeanMapperBuilder<JavaBeanMappingHelper> create() {
        return new JavaBeanMapperBuilder<JavaBeanMappingHelper>();
    }

    public JavaBeanMapperBuilder<H> withValueMapping(AValueMappingDef<?,?,? super H> m) {
        valueMappings.add(m);
        return this;
    }

    public JavaBeanMapperBuilder<H> withObjectMapping(AObjectMappingDef<?,?,? super H> m) {
        objectMappings.add(m);
        return this;
    }

    @SuppressWarnings("unchecked")
    public JavaBeanMapperBuilder<H> withBeanMapping(JavaBeanMapping m) {
        objectMappings.add(m.build());

        if(m.getSourceClass() != m.getTargetClass()) {
            objectMappings.add(m.buildBackward());
        }
        return this;
    }


    public JavaBeanMapperBuilder<H> withLogger(AMapperLogger logger) {
        this.logger = logger;
        return this;
    }

    public JavaBeanMapperBuilder<H> withIdentifierExtractor(AIdentifierExtractor ie) {
        this.identifierExtractor = ie;
        return this;
    }

    public JavaBeanMapperBuilder<H> withHelperFactory(AFunction0<H, RuntimeException> helperFactory) {
        this.helperFactory = helperFactory;
        return this;
    }

    public JavaBeanMapperBuilder<H> withHelper(final H helper) {
        this.helperFactory = new AFunction0<H, RuntimeException>() {
            @Override public H apply() {
                return helper;
            }
        };
        return this;
    }

    public JavaBeanMapperBuilder<H> withContextExtractor(AContextExtractor contextExtractor) {
        this.contextExtractor = contextExtractor;
        return this;
    }

    public JavaBeanMapperBuilder<H> withPreProcessor(APreProcessor preProcessor) {
        this.preProcessors.add(preProcessor);
        return this;
    }

    public JavaBeanMapperBuilder<H> withPostProcessor(APostProcessor postProcessor) {
        this.postProcessors.add(postProcessor);
        return this;
    }

    public JavaBeanMapper build() {
        return build(false);
    }

    @SuppressWarnings("unchecked")
    public JavaBeanMapper build(boolean compile) {
        Collections.reverse(objectMappings);
        Collections.reverse(valueMappings);

        try {
            return new JavaBeanMapperImpl<H> (
                    objectMappings, valueMappings,
                    logger, helperFactory, identifierExtractor, contextExtractor,
                    preProcessors, postProcessors, compile
            );
        } catch (Exception exc) {
            return AMapperExceptionHandler.onError(exc, APath.EMPTY);
        }
    }
}
